### PR TITLE
client: Drop two calls to `pretty.ToJSON` and move code outside of lock

### PR DESCRIPTION
**Pull request by townba**
RELEASE NOTES:
* client: improve RPC performance by reducing work while holding a lock

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -934,10 +933,14 @@ func equalAddresses(a, b []resolver.Address) bool {
...
 func (ac *addrConn) updateAddrs(addrs []resolver.Address) {
-	ac.mu.Lock()
-	channelz.Infof(logger, ac.channelz, "addrConn: updateAddrs curAddr: %v, addrs: %v", pretty.ToJSON(ac.curAddr), pretty.ToJSON(addrs))
-
 	addrs = copyAddressesWithoutBalancerAttributes(addrs)
+	limit := len(addrs)
+	if limit > 5 {
+		limit = 5
+	}
+	channelz.Infof(logger, ac.channelz, "addrConn: updateAddrs addrs (%d of %d): %v", limit, len(addrs), addrs[:limit])
```
Comment (dfawley) [inline clientconn.go:941]: Can you show what this looks like, please?  (Mainly I'm not sure: does `%v` on a slice call `String()` for each member?)

**Review by dfawley — COMMENTED** [review]
Could you please add a comment to `addrConn`'s `mu` field that says something like "`This is taken on the RPC path, so its usage should be minimized as much as possible.  TODO: find a lock-free way to retrieve the transport and state from the addrConn without a lock`"

File: clientconn.go (hunk 1 above)
Comment (townba) [reply clientconn.go:941]: Looks something like this:
```
addrConn: updateAddrs addrs (3 of 3): [{Addr: "127.0.0.1:35771", ServerName: "", } {Addr: "127.0.0.1:41729", ServerName: "", } {Addr: "127.0.0.1:36587", ServerName: "", }]
```

**Comment by townba** [comment]
> Could you please add a comment to `addrConn`'s `mu` field that says something like …

Done.

**Comment by townba** [comment]
PTAL

File: clientconn.go (hunk 1 above)
Comment (dfawley) [reply clientconn.go:941]: Great, thanks!

File: clientconn.go [hunk 2]
```diff
@@ -1172,6 +1175,10 @@ type addrConn struct {
 	// is received, transport is closed, ac has been torn down).
 	transport transport.ClientTransport // The current transport.
 
+	// This mutex is used on the RPC path, so its usage should be minimized as
+	// much as possible.
+	// TODO: Find a lock-free way to retrieve the transport and state from the
+	// addrConn without a lock.
```
Comment (dfawley) [inline clientconn.go:1181]: Oops, "lock-free ... without a look" - please remove one or the other.  Thanks!

**Review by dfawley — APPROVED** [review]

File: clientconn.go (hunk 2 above)
Comment (townba) [reply clientconn.go]: Done. PTAL

**Review by dfawley — APPROVED** [review]

**Pull request merged**